### DISCUSSION TITLE
SSG page data json files respond with rendered page html

### DISCUSSION THREAD

**Discussion by gijsbotje** [vercel/vercel#8372] [Help] [answered]
In my latest deployments of https://dbm-nextjs.vercel.app/ all SSG pages work on first load, but client side routing breaks. This is because for these pages, the JSON file that should contain the pageProps, somehow return the rendered page as HTML.

For example: https://dbm-nextjs.vercel.app/blog renders fine
but when you click on a blog and navigate back, you'll end up on a page with only the layout and no data.
It should load the pageProps from https://dbm-nextjs.vercel.app/_next/data/p68BYzThr7N3EAjTTvZgA/nl-nl/blog.json, but this returns HTML.

How is this possible? 6 days ago it worked fine: https://dbm-nextjs-n7ru7lzx6-afosto.vercel.app/_next/data/l3byLPuS4BLk5D5N0cHjS/nl-nl/blog.json
now it doesn't: https://dbm-nextjs.vercel.app/_next/data/p68BYzThr7N3EAjTTvZgA/nl-nl/blog.json

Running a local build with `yarn start` works fine. It's only the deployments online that are broken.

**Comment by amyegan** [comment]
Do you have a specific version of Next.js pinned in your project, or is it possible that your local version is not the same as what Vercel is using to build + deploy? Or have you made any other changes, even things that seem unrelated, to your code or project configuration around the time deployments started erroring? Those are the first things I like to check when builds suddenly break. 

If you can share your repo or a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), I would be happy to take a closer look.

**Reply by gijsbotje** [reply]
The first build that started going wrong only had some minor tweaks to components, nothing changed in any getServerSideProps. 
I did add a reportWebVitals function to _app, but that is actually the last build that worked.
There are 2 other pages (dynamic pages) that had no changes at all that also return HTML instead of JSON.

**Collection pages**
https://dbm-nextjs.vercel.app/_next/data/p68BYzThr7N3EAjTTvZgA/nl-nl/ringen/bruine-diamant.json
should be 
https://dbm-nextjs-n7ru7lzx6-afosto.vercel.app/_next/data/l3byLPuS4BLk5D5N0cHjS/nl-nl/ringen/bruine-diamant.json


**Product pages**
https://dbm-nextjs.vercel.app/_next/data/p68BYzThr7N3EAjTTvZgA/nl-nl/ringen/witgouden-alliance-aanschuifring-ronde-vorm-diamant-moderne-jackie-half.htm.json
should be
https://dbm-nextjs-n7ru7lzx6-afosto.vercel.app/_next/data/l3byLPuS4BLk5D5N0cHjS/nl-nl/ringen/witgouden-alliance-aanschuifring-ronde-vorm-bruine-diamant-moderne-jackie-half.htm.json

**Reply by amyegan** [reply]
Thanks for sharing these examples. Are you able to share your repo or a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example)? I see that the latest deployment is returning a 404 for the addresses you provided, but I can't see why that's happening from the browser information alone.

<img width="497" alt="image" src="https://user-images.githubusercontent.com/8247105/184383043-ac6a9267-d153-4616-820e-52fb316219a2.png">

Rough ideas of what could cause a 404 on data fetching:
- You may need to toggle [Directory Listing](https://vercel.com/docs/concepts/edge-network/directory-listing)
- An uncaught error might be coming back from your data source (could be related to expired or missing token in env vars)
- Something could be off in your `vercel.json` or `next.config.json`
- Vercel may be using a newer version of one of your installed packages than what is installed on your local computer (depending on your `packages.json` and lockfile)

If you can share a minimal reproducible example, or more details about your project, then I can dig into this issue a bit more.

**Reply by gijsbotje** [reply]
@amyegan Sorry, that build was overwritten. This is a comparison that will stay:
**Collection pages**
https://dbm-nextjs-gph0vhql3-afosto.vercel.app/_next/data/pACGhGClEtyXLM01OaMop/nl-nl/collection/ringen/bruine-diamant.json
should be 
https://dbm-nextjs-n7ru7lzx6-afosto.vercel.app/_next/data/l3byLPuS4BLk5D5N0cHjS/nl-nl/ringen/bruine-diamant.json


I tried making a test case repo, but it doesn't show the same behavior.

We do use middleware, but that only rewrites the URL based on a type we get from our API. And the blog is ignored for this, and the index page (which is SSR) also returns HTML for the JSON request.

https://dbm-nextjs-gph0vhql3-afosto.vercel.app/_next/data/pACGhGClEtyXLM01OaMop/nl-nl/blog.json

The next config file looks like this:
```
module.exports = {
  reactStrictMode: true,
  swcMinify: true,
  compiler: {
    emotion: true,
  },
  async rewrites() {
    return {
      beforeFiles: [
        {
          source: '/kennisbank/:path*',
          destination: '/education-center/:path*',
        },
        {
          source: '/datenbank/:path*',
          destination: '/education-center/:path*',
        },
        {
          source: '/centre-de-connaissance/:path*',
          destination: '/education-center/:path*',
        },
        {
          source: '/klantenservice/:path*',
          destination: '/customer-service/:path*',
        },
        {
          source: '/kundenservice/:path*',
          destination: '/customer-service/:path*',
        },
        {
          source: '/service-clients/:path*',
          destination: '/customer-service/:path*',
        },
      ],
    };
  },
  images: {
    domains: [
      'source.unsplash.com',
      'images.unsplash.com',
      'images.prismic.io',
      'static.diamondsbyme.com',
      'cdn2.diamondsbyme.com',
    ],
  },
  i18n: {
    locales: ['en-gb', 'nl-nl', 'nl-be', 'en-us', 'fr-fr', 'de-de'],
    defaultLocale: 'nl-nl',
    localeDetection: false,
  },
};
```

**Reply by amyegan** [reply]
I don't see anything in your config that should cause this behavior. I have seen HTML returned when an error comes back from the underlying service in the past. For example, when pulling data from CMS and a 404 or other error comes back and the app mishandles the error. 

Without a code reproduction it's difficult to say exactly what is causing it. Do you have any [logging](https://vercel.com/docs/concepts/deployments/logs) in your function that could help you locate the the exact point where things are going off course? The guides on [Where can I find my logs on Vercel?](https://vercel.com/guides/where-are-vercel-logs) and [How do I store logs on Vercel?](https://vercel.com/guides/how-do-i-store-logs-on-vercel) can help if you don't already have some logging set up.

**Comment by PhilippSchreiber** [comment]
I see exactly the same behavior on my page since I added a middleware and upgraded to the newest next version. But only on vercel. Locally everything works fine. Here is a link where you can see the wrong behavior:

https://kruucom-bojvfha80-kruu.vercel.app/

In the network tab when I search for "de.json" I see a rendered HTML. When I open the file directly it returns the correct json.

Unfortunately my tries to create a minimum example where not successful.

**Comment by PhilippSchreiber** [comment]
@gijsbotje for me it looks like the issue is solved after upgrading to 12.3. Can you confirm?

**Reply by gijsbotje** [reply]
@PhilippSchreiber Unfortunately it's still happening on 12.3.0
https://dbm-nextjs-git-feature-update-packages-afosto.vercel.app/blog
click on an article -> navigate back

**Reply by PhilippSchreiber** [reply]
OK, unfortunately it is still present in 12.3.0

**Comment by PhilippSchreiber** [comment]
@gijsbotje we upgraded to next 13 and deployed it to live. This time it really looks, like its gone. Can you confirm?

**Comment by gijsbotje** [comment, marked as answer]
One of my colleagues did a performance audit and fixed some bugs. We are not sure how or what he fixed, but it resolved the issue.

**Reply by PhilippSchreiber** [reply]
Yeah, fixed by accident 😄

**Reply by gijsbotje** [reply]
@PhilippSchreiber exactly, a case of "My code works, I have no idea why"

**Comment by japboy** [comment]
i have same issue with `next@10.2.0`, `next@12.3.1`, or `next@12.3.4` on the vercel environment. i haven't checked with `next@13.x` though...

@gijsbotje any guesses what actually have solved your problem?

**Reply by gijsbotje** [reply]
I believe our fix came from restructuring the data that was returned. It might also haven been a broken component, but not sure that was it.
You could try debugging it by removing al returned props and test if it works, then gradually adding more data to identify the data thats breaking the page. Same goes for compoments if the problem isn't the data.

**Reply by japboy** [reply]
thanks for reply, and i found why this happens to me. in my case, there were some [URL rewrite rules](https://vercel.com/docs/project-configuration#legacy/routes) in `vercel.json` and the fallback value was 404 error page. this caused the same issue. idk for others though...
